README.md
# DMC with Parallel Computation and C++ Capabilities

ggdmc implements hierarchical Bayesian, evidence accumulation model (HB-EAM). 
This release includes drift-diffusion model (Ratcliff, 1978). Based on 
Voss, Rothermund, and Voss's (2004) fast-dm 30.2 density.c, the DDM in
ggdmc implements the equations in their paper as C++ routines plus a new
parallel integration function to handle high precision requirement. 

ggdmc is derived from Andrew Heathcote's Dynamic Model of Choice (DMC), 
which has also implemented other numerous EAMs.  Identical to DMC, ggdmc uses 
differential evolution Markov Chain Monte Carlo sampler to search optimal 
theta and phi that maximise posterior likelihood. 

## Getting Started
Here is a simple example extracted from Andrew Heathcote's DMC workshop 
materials. For further details, please see R help pages in this package. 

```
require(ggdmc) 

## Use a 6-parameter drift-diffusion model  
## The only experimental factor is Stimulus type
m1 <- model.dmc(
  p.map     = list(a="1", v="1", z="1", d="1", sz="1", sv="1", t0="1",st0="1"),
  constants = c(st0=0, d=0),          ## Fixed st0 and d at 0
  match.map = list(M = list(s1 = "r1", s2 = "r2")),
  factors   = list(S=c("s1", "s2")),  ## Two stimulus types  
  responses = c("r1","r2"),           ## Two response types
  type      = "rd")                   ## rd stands for Ratcliff's diffusion

## Set up a prior probability distribution setting
p.prior <- prior.p.dmc(
  dists = rep("tnorm", 6),
  p1    = c(a=2,  v=2.5, z=.5, sz=.3, sv=1,  t0=.3),
  p2    = c(a=.5, v=.5,  z=.1, sz=.1, sv=.3, t0=.05),
  lower = c(0,-5, 0, 0, 0, 0),
  upper = c(5, 7, 2, 2, 2, 2))

## Assume a true DDM parameter vector
p.vector <- c(a=1, v=1, z=.5, sz=.25, sv=0.2,t0=.15)

## Use simulate function to simulate choice-RT data 
## One usually would like to fit his/her own empirical data.
dat1 <- simulate(m1, nsim=1e2, p.vector=p.vector)

## Set up a data model instance. This binds the empirical/simulated data with 
## the model set-up
mdi1 <- data.model.dmc(dat1, m1)

## Initialise a small sample 
## (1) iteration number == 250 
## (2) thinning length == 1
## (3) prior distributions are listed in p.prior 
## (4) data == mdi1, an assumed model and a simulated/empirical data frame
samples0 <- samples.dmc(nmc=250, p.prior=p.prior, data=mdi1, thin=1)

## Fit the Bayesian model 
samples0 <- run.dmc(samples0)


## Check if model converged, etc.
gelman.diag.dmc(samples0)
## Potential scale reduction factors:
## 
##    Point est. Upper C.I.
## a        1.16       1.29
## v        1.20       1.36
## z        1.18       1.33
## sz       1.12       1.21
## sv       1.24       1.48
## t0       1.16       1.34

plot(samples0)  ## Check traceplot to see if chains converged

## Further details and more arguments, please see
?gelman.diag.dmc
?samples.dmc
?run.dmc
?plot.dmc

## For hierarchical Bayesian model 
?h.run.dmc
?h.samples.dmc
?plot.dmc.list
?plot.hyper

```


### Prerequisites
 - R (>= 3.0.2)
 - Rcpp (>= 0.12.3), RcppArmadillo (>= 0.6.700.6.0), ggplot2 (>= 2.1.0),
   rtdists (>= 0.6-6), gridExtra (>= 2.2-1), ggmcmc (>= 0.7.3), 
   ggthemes (>= 3.0.1), stats (>= 3.2.2), loo (>= 0.1.6), coda (>= 0.16-1)
 - Windows users need Rtools (>= 3.3.0.1959), and Microsoft Visual Studio 
   Community (>= 2015) (for Open MPI library and M_PI macro support)
 - OS X users need to install Open MPI library
 - Linux/Unix users may need to install Open MPI library, if it has not 
   been installed. 
 - [Armadillo](https://CRAN.R-project.org/package=RcppArmadillo)
   requires a recent compiler; for the g++ family at least version 4.6.*
   is required. Some Linux/Unix distributions may use a g++ earlier than 4.6.*.  
 - This package uses Rcpp::Nullable, so Rcpp >= 0.12.1 is required.
   

Successful cases for Windows OS:
  - Microsoft Visual Studio Community 2015 (Version 14.0.25421.03 Update 3) on  
    Windows 10 64 bits.
  - Microsoft Visual Studio Community 2015 (Version 14.0.24720.1 Update 1), 
    with Rtools 3.4 on Windows 10 64 bits.
  
Unsuccseeful cases for Windows OS:
  - Microsoft Blend for Visual Studio Express 2015   

### Installing

```
From CRAN: install.packages("ggdmc")
From source: install.packages("ggdmc_0.1.3.5.tar.gz", repos = NULL, type="source")

### Other supporting packages for DMC (not necessary for ggdmc)
install.packages("coda_0.18-1.3.tar.gz", repos = NULL, type="source")
install.packages("tnorm_0.1.0.0.tar.gz", repos = NULL, type="source")

```

## Citation

If you use this package, please cite the software, for example:

Lin, Y.-S., & Heathcote, A (in preparation). ggdmc: An R package for 
hierarchical Bayesian evidence accumulation models, using differential
evolution Markov Chain Monte Carlo Sampler. Retrieved from
https://github.com/TasCL/ggdmc

## Contributors

The ggdmc C++ codes are developed by [Yi-Shin Lin](http://www.tascl.org/yi-shin-lin.html). 

The R codes mainly are incorporated from DMC developed by [Andrew Heathcote](http://www.tascl.org/andrew-heathcote.html). 

If there is any bug been introduced inadvertently into DMC R codes, they are 
probably errors brought in by the first author. Please report any bugs you may 
find to the first [author](mailto:yishin.lin@utas.edu.au). 

## License

GPL-2. Please see License.md/LICENSE for details.

## Acknowledgments

* density.cpp is based on Voss & Voss's (2012) density.c in fast-dm 30.2. Any 
identical parts in density.cpp belong to their copyright.
* The tnorm_wrapper.cpp and tnorm.R are based on Jonathan Olmsted's
<jpolmsted@gmail.com> RcppTN 0.1-8 (https://github.com/olmjo/RcppTN) 
and Christopher Jackson's <chris.jackson@mrc-bsu.cam.ac.uk> R codes in msm. 
Any identical parts in tnorm_wrapper.cpp and tnrom.R belong to their copyright.
* C++ codes in this package depend largely on Dirk Eddelbuettel, Romain 
Francois and Doug Bates's Rcpp, and RcppArmadillo.  
* Armadillo is a collection of C++ library for performing linear
algebra <http://arma.sourceforge.net/>, authored by Conrad Sanderson. 
* DEMCMC sampler is based on the following papers, Turner & Sederberg (2012),
Turner et al (2013) and Ter Braak (2006).
* Thanks to Matthew Gretton's consulation about rtdists's internal. 


NEWS
0.1.3.6 Yi-Shin Lin <yishin.lin@utas.edu.au>

	- Correct 'overloading ambiguity" installation failures on Solaris

0.1.3.5 Yi-Shin Lin <yishin.lin@utas.edu.au>

	- First release to the public

0.1.3.4 Yi-Shin Lin <yishin.lin@utas.edu.au>

	- improve ddmc.Rd, dmc.Rd, dprior.Rd.
	- add dmc.R to generate ggdmc-package.Rd
	- merge DMC's and ggdmc's plot.prior and rename drawPrior as plot.priors
	- now ggdmc has its own classes dmc, dmc.list and hyper
	- ggdmc now uses S3 class to dispatch different summary and plot functions.
	- Finally, LBA model now is in ggdmc. likelihood.dmc and transform.dmc branch
	out to e.g., logLik.ddm, logLik.lba_B, becasue two new classes, ddm and
	lba_B are added.
	- New modules ggdmc_class.R and ggdmc_eam.R are added. The former defines
	DMC classes and the latter collects evidence accumulator models.
	- Add more help pages.
	- profile.dmc merge with stats's profile
	- roxygen2 now is also in charge of NAMESPACE

0.1.3.0 Yi-Shin Lin <yishin.lin@utas.edu.au>

	- plot.dmc add function to superimpose prior distribution on top of
	posterior distribution (ggplot2).
	- DMC needs to call modified coda to use this function (base).


src/utility.cpp
//   Copyright (C) <2016>  <Yi-Shin Lin>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation; version 2
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License along
//   with this program; if not, write to the Free Software Foundation, Inc.,
//   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#include <RcppArmadillo.h>
#include <ggdmc.hpp>

Rcpp::NumericMatrix na_matrix(int nr, int nc) {
  Rcpp::NumericMatrix m(nr, nc) ;
  std::fill( m.begin(), m.end(), Rcpp::NumericVector::get_na() ) ;
  return m ;
}

arma::uvec getArmaIdx(Rcpp::NumericVector x, int trueOrFalse) {
  arma::vec xArma(x.begin(), x.size(), false) ;
  arma::uvec index = find(xArma == trueOrFalse) ;
  return index ;
}

/* This function uses 1st subject's p.prior as a template and replaces
 the location and scale for each parameter with those generated from
hyper distributions. The return is a nChains list with each has its own
p.prior */
// [[Rcpp::export]]
Rcpp::List assign_pp_pLists(Rcpp::List& samples_s1, Rcpp::List& usePhi) {
  Rcpp::List pList         = samples_s1["p.prior"] ;
  int nChains              = samples_s1["n.chains"] ;
  arma::mat usePhiLocation = usePhi[0] ; // nChains x npars
  arma::mat usePhiScale    = usePhi[1] ;
  Rcpp::List pLists(nChains) ; //p.priors (plural)
  Rcpp::List dist_setting ;

  // let usePhi follow normal order eg 0, 1, ..., 23
  // But shuffle the chain sequence
  Rcpp::IntegerVector chainSeq = Rcpp::seq_len(nChains) ; /* eg 1:24 */

  // obtain a time-based seed:
  unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
  std::shuffle(chainSeq.begin(), chainSeq.end(), std::default_random_engine(seed));

  for(int i =0; i<nChains; i++)   // eg 0 to 23
  {
    int ii = chainSeq[i] - 1 ;
    Rcpp::List pList_cp(clone(pList)) ; // Make hard copies several times
    for(Rcpp::List::iterator it2=pList_cp.begin(); it2!=pList_cp.end(); ++it2)
    {
      int j = std::distance(pList_cp.begin(), it2) ; // j is ordered
      dist_setting    = *it2 ; // pList_cp's memory location is altered
      dist_setting[0] = as_scalar(usePhiLocation.row(i).col(j)) ;
      dist_setting[1] = as_scalar(usePhiScale.row(i).col(j)) ;
    }
    pLists[ii] = pList_cp ; // ii is random pList_cp follows the order
  }
return pLists ;
}

/* been used in summed_log_likelihood_hyper. This function changes
 * memory location directly; input pList will be modified */
void assign_pp_pList(Rcpp::List& phi, Rcpp::List& pList) {
  // To replace 1st assign.pp; locat and scale has size of npar
  arma::vec location = phi[0] ;
  arma::vec scale    = phi[1] ;
  for(Rcpp::List::iterator it = pList.begin(); it != pList.end(); ++it)
  { // Loop through a, v, z, sv, ...
    int i = std::distance(pList.begin(), it) ;
    Rcpp::List pList_same_mem_loc = *it ;
    pList_same_mem_loc[0] = location(i) ;
    pList_same_mem_loc[1] = scale(i) ;
  }
}

Rcpp::List swap(Rcpp::NumericVector& pVec, Rcpp::List& pList) {
  std::vector<std::string> pVecNames  = pVec.names() ;
  std::vector<std::string> pListNames = pList.names() ;
  int pVecSize  = pVec.size() ;
  int pListSize = pList.size() ;
  Rcpp::List out(pList.size()) ;

  if (pVecSize != pListSize)
  {
    Rcpp::stop("p.vector does not match parameter prior list.") ;
  }

  for (std::vector<std::string>::iterator it1 = pVecNames.begin();
    it1 != pVecNames.end(); ++it1)
  {
    int i = std::distance(pVecNames.begin(), it1) ;
    if (pVecNames[i] != pListNames[i])
    {
      Rcpp::Rcout << "Found an element in p.prior and p.vector is ordered "
                  << "differently. p.prior will be rearranged to follow "
                  << "p.vector.\n" ;
      for (Rcpp::List::iterator it2 = pList.begin(); it2 != pList.end(); ++it2)
      {
        int j = std::distance(pList.begin(), it2) ;
        if (pListNames[j] == pVecNames[i])
        {
          out[i] = pList[j];
        }
      }
    } else
    {
      out[i] = pList[i];
    }
  }

  out.names() = pVecNames ;
  return out ;
}

Rcpp::List isConstant(Rcpp::List& ppList) {
  Rcpp::List location = ppList[0] ;
  Rcpp::List scale    = ppList[1] ;
  Rcpp::LogicalVector location_out(location.size()) ;
  Rcpp::LogicalVector scale_out(scale.size()) ;
  for(int i=0; i < location.size(); i++)
  {
    Rcpp::List a = location[i] ;
    std::string b = a.attr("dist") ;
    location_out[i] = b == "constant" ;
  }

  for(int j=0; j < scale.size(); j++)
  {
    Rcpp::List a = scale[j] ;
    std::string b = a.attr("dist") ;
    scale_out[j] = b == "constant" ;
  }

  Rcpp::List out = Rcpp::List::create(
    Rcpp::Named("location") = location_out,
    Rcpp::Named("scale")    = scale_out) ;

  return out;
}

Rcpp::List resize( const Rcpp::List& x, int n ) {
  int oldsize = x.size() ;
  Rcpp::List y(n) ;
  for( int i=0; i<oldsize; i++) y[i] = x[i] ;
  return y ;
}

arma::cube get_ps(Rcpp::List& samples) {
  /* Collect 1st nmc and reshape ps cube as nSubjects x npars x nChains
   * Andrew's ps array is nChains x nSubjects x npars; but Armadillo's cube
   * uses slice order first, so chain dimension has to be in slice */
  Rcpp::List hyper = samples.attr("hyper") ;
  int nChains      = hyper["n.chains"] ;
  int nsubs        = samples.size() ;
  int npars        = hyper["n.pars"] ;
  int start_R      = hyper["start"] ; // Note start is R index

  arma::cube ps(nsubs, npars, nChains) ;
  for(Rcpp::List::iterator it = samples.begin(); it != samples.end(); ++it)
  {
    int it_i = std::distance(samples.begin(), it) ;
    Rcpp::List subject_i = *it ;
    arma::cube theta_s   = subject_i["theta"] ;        // nChain x npars x nmc
    arma::mat theta_nmc1 = theta_s.slice(start_R - 1); // take 1st nmc; nChain x npars
    for(int i=0; i<nChains; i++)
    {
      ps.slice(i).row(it_i) = theta_nmc1.row(i) ; // nsubs x npars x nchains
    }
  }
  return ps ;
}

arma::cube get_ps_use(Rcpp::List samples) {
  Rcpp::List samples1 = samples[0] ;
  Rcpp::List use1     = samples1.attr("use") ;
  arma::mat theta1    = use1["theta"] ;
  int nChains         = theta1.n_rows ;
  int nsubs           = samples.size() ;
  int npars           = theta1.n_cols ;

  arma::cube ps(nsubs, npars, nChains) ;
  for(Rcpp::List::iterator it = samples.begin(); it != samples.end(); ++it)
  {
    int it_i = std::distance(samples.begin(), it) ;
    Rcpp::List subject_i = *it ;
    Rcpp::List use       = subject_i.attr("use") ;
    arma::mat theta_i    = use["theta"] ;        // nChain x npars
    for(int i=0; i<nChains; i++)
    { // nsubs x npars x nchains
      ps.slice(i).row(it_i) = theta_i.row(i) ;
    }
  }
  return ps ;
}

void add_nmc0 (Rcpp::List& samples, Rcpp::List& pLists) {
  for(Rcpp::List::iterator it=samples.begin(); it!=samples.end(); ++it)
  {
    int it_i = std::distance(samples.begin(), it) ;
    Rcpp::List samples_i  = *it ;
    arma::cube theta_i    = samples_i["theta"] ; // nChains x npars x nmc
    arma::mat logprior_i  = samples_i["summed_log_prior"] ; // nmc x nChains
    arma::mat loglike_i   = samples_i["log_likelihoods"] ;  // nmc x nChains
    int start_R  = samples_i["start"] ; // R index
    int start_C  = start_R - 1 ;        // C index
    arma::mat theta    = theta_i.slice(start_C) ; // nChains x npars
    arma::vec logprior = vectorise(logprior_i.row(start_C)) ;   // nChains x 1
    arma::vec loglike  = vectorise(loglike_i.row(start_C)) ;   // nChains x 1
    Rcpp::List use = Rcpp::List::create(
      Rcpp::Named("theta")    = theta,    // nChains x npars
      Rcpp::Named("logprior") = logprior, // nChains x 1
      Rcpp::Named("loglike")  = loglike,  // nChains x 1
      Rcpp::Named("store_i")  = start_C,
      Rcpp::Named("p.priors") = pLists) ;
    samples_i.attr("use") = use ;
    samples[it_i] = samples_i ;
  }
}

//' Sum and Log Probability Density of a EAM model
//'
//' Get log likelihood summed over a model data instance.  The input data has
//' to be a data frame carrying a model specification, which is usually
//' created by \code{model.data.dmc} function.
//' \code{summed_log_likelihood_parallel} does calculation in parallel. Use it
//' when only the data set is big.
//'
//' @param pVec a parameter vector
//' @param data a model data instance
//' argument.
//' @return a double scalar
//' @export
//' @examples
//' m1 <- model.dmc(
//'   p.map     = list(a="1", v="1", z="1", d="1", sz="1", sv="1", t0="1", st0="1"),
//'   constants = c(st0=0, d=0),
//'   match.map = list(M=list(s1="r1", s2="r2")),
//'   factors   = list(S=c("s1", "s2")),
//'   responses = c("r1", "r2"),
//'   type      = "rd")
//'
//' pVec <- c(a=1, v=1, z=0.5, sz=0.25, sv=0.2,t0=.15)
//'
//' ## Set up a model-data instance
//' dat <- simulate(m1, nsim=1e2, p.vector=pVec)
//' mdi <- data.model.dmc(dat, m1)
//' summed_log_likelihood(pVec, mdi)
//' ## [1] 0.3796048
// [[Rcpp::export]]
double summed_log_likelihood(arma::vec& pVec, Rcpp::List& data) {
    // TODO: check pVec_NV and pVecNA names match
    // Should find Rcpp sugar match
  Rcpp::NumericVector model    = data.attr("model") ;
  Rcpp::NumericVector pVecNA   = model.attr("p.vector") ;
  Rcpp::CharacterVector pNames = pVecNA.names() ;
  Rcpp::NumericVector pVec_NV  = Rcpp::as<Rcpp::NumericVector>(Rcpp::wrap(pVec)) ;
  pVec_NV.names()              = pNames ;
  std::vector<double> density  = ddmc(data, pVec_NV, 2.5, 1e-10) ;
  double summedLogLike         = 0 ;

  for(std::vector<double>::iterator it=density.begin(); it!=density.end();
  ++it)
  { summedLogLike += log(*it) ; }
  return summedLogLike ;
}


//' @rdname summed_log_likelihood
//' @export
// [[Rcpp::export]]
double summed_log_likelihood_parallel(arma::vec& pVec, Rcpp::List& data) {
  Rcpp::NumericVector model    = data.attr("model") ;
  Rcpp::NumericVector pVecNA   = model.attr("p.vector") ;
  Rcpp::CharacterVector pNames = pVecNA.names() ;
  Rcpp::NumericVector pVec_NV  = Rcpp::as<Rcpp::NumericVector>(Rcpp::wrap(pVec)) ;
  pVec_NV.names()              = pNames ;
  double summedLogLike         = 0 ;

  std::vector<double> density = ddmc_parallel(data, pVec_NV, 2.5, 1e-10) ;

  for(std::vector<double>::iterator it=density.begin(); it!=density.end();
  ++it)
  {
    summedLogLike += log(*it) ;
  }
  return summedLogLike ;
}

double summed_log_likelihood_hyper(arma::mat& data_hyper,
  Rcpp::List& pList_hyper) {
  /* hyper log likelihood sums over all DDM parameters and over all
  participants replacing the location and scale parameters for each
  DDM parameter with the updated phi vectors
  * data_hyper is from data_hyper_ii, one chain from migrate subchains */
  Rcpp::CharacterVector pNames = pList_hyper.names() ;
  int nSubjects = data_hyper.n_rows ; // (nSubjects x npars)

  double sumOverSubjects = 0;
  for(int i=0; i<nSubjects; i++)
  {
    arma::mat pVec_i = data_hyper.row(i) ; // 1 x npars mat
    Rcpp::NumericVector pVecNV_i =
      Rcpp::as<Rcpp::NumericVector>(Rcpp::wrap(pVec_i)) ;
    pVecNV_i.names() = pNames ;
    double sumOverPars = sum(dprior(pVecNV_i, pList_hyper)) ;
    sumOverSubjects += sumOverPars ;
  }

  return sumOverSubjects ;
}


//' Sum and Log Prior Density of a EAM model
//'
//' Get log likelihood summed over all prior parameters
//'
//' @param pVec a parameter vector
//' @param pPrior p.prior
//' @export
//' @examples
//' pVec <- c(a=1, v=1, z=0.5, sz=0.25, sv=0.2,t0=.15)
//'
//' p.prior  <- prior.p.dmc(
//'   dists = rep("tnorm", 6),
//'   p1    = c(a=2,   v=2.5, z=0.5, sz=0.3, sv=1,  t0=0.3),
//'   p2    = c(a=0.5, v=.5,  z=0.1, sz=0.1, sv=.3, t0=0.05) * 5,
//'   lower = c(0,-5, 0, 0, 0, 0),
//'   upper = c(5, 7, 2, 2, 2, 2))
//'
//' summed_log_prior(pVec, p.prior)
//' ## -3.224406
// [[Rcpp::export]]
double summed_log_prior(arma::vec& pVec, Rcpp::List& pPrior) {
  Rcpp::NumericVector pVec_NV  = Rcpp::as<Rcpp::NumericVector>(Rcpp::wrap(pVec)) ;
  std::vector<std::string> pNames = pPrior.names() ;
  pVec_NV.names() = pNames ;
  return sum(dprior(pVec_NV, pPrior)) ;
}

double summed_log_prior_hyper(Rcpp::List& pVec, Rcpp::List& ppList) {
  arma::mat phiVecLocation = pVec[0] ;
  arma::mat phiVecScale    = pVec[1] ;
  Rcpp::NumericVector location =
    Rcpp::as<Rcpp::NumericVector>(Rcpp::wrap(phiVecLocation)) ;
  Rcpp::NumericVector scale    =
    Rcpp::as<Rcpp::NumericVector>(Rcpp::wrap(phiVecScale)) ;
  Rcpp::List ppList_location            = ppList[0] ;
  Rcpp::List ppList_scale               = ppList[1] ;
  Rcpp::CharacterVector pNames_location = ppList_location.names() ;
  Rcpp::CharacterVector pNames_scale    = ppList_scale.names() ;
  location.names() = pNames_location ;
  scale.names()    = pNames_scale ;
  double out = sum(dprior(location, ppList_location)) +
    sum(dprior(scale, ppList_scale)) ;
  return out ;
}

arma::vec update_useLogPrior(arma::mat& theta, Rcpp::List& pLists)
{
  int nChains = theta.n_rows ;   // nChains x npars
  arma::vec out(nChains) ;
  for(int i=0; i<nChains; i++)
  {
    arma::vec pVec   = vectorise(theta.row(i)) ;
    Rcpp::List pList = pLists[i] ; // new priors
    out(i)           = summed_log_prior(pVec, pList) ;
  }
  return out ;
}

